### PR TITLE
vlc: 2.2.8 -> 3.0.2

### PR DESCRIPTION
**Pull request by vaibhavsagar**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by vaibhavsagar** [comment]
Currently getting

```
/nix/store/adidfx4pa7vmvby0gjqqmiwg2x49yr27-binutils-2.28.1/bin/ld: cannot find -lgpg-error
```

halfway through compilation.

**Comment by vaibhavsagar** [comment]
Now I'm having issues with the `installPhase`.

**Comment by fpletz** [comment]
@GrahamcOfBorg build vlc

**Comment by pbogdan** [comment]
@vaibhavsagar try adding `gettext` to `nativeBuildInputs`. I also needed `qtsvg` to get the qt5 GUI building but maybe that can be avoided somehow? Not sure.. Oh, and 3.0.1 was released in the mean time as well.

**Comment by pbogdan** [comment]
Two more notes from looking at this bit more:

1. On startup vlc would print a bunch of warnings about stale plugin cache entries:

    ```
    stale plugins cache: modified /nix/store/yzbyk4c8m3xsz39y47akpbn013lzxayy-vlc-3.0.1/lib/vlc/plugins/misc/libexport_plugin.so
    ```

    It seems that it's maintaining some sort of cache / database of those plugins and from a quick look the validity of the entries is based on the mtime of the file as well as its size. Since the cache looks to be generated in the install phase it's problematic as Nix will reset the mtimes and I guess the fixup phase may change the sizes too? (with stripping / shrinking etc.).
    I worked around it for now with an ugly hack:

    ```
    postFixup = ''
      find $out/lib/vlc/plugins -exec touch -d @1 '{}' ';'
      $out/lib/vlc/vlc-cache-gen $out/vlc/plugins
    '';
    ```
    which resets the mtimes to what they will be in the store, does so after the fixup phase and regenerates the cache. There is probably a better way to work around this but thought I would at least mention this.
2. `phonon-backend-vlc` fails to build with vlc 3 but an upstream update of that package is available and appears to fix the build failure.

**Comment by vaibhavsagar** [comment]
Thanks @pbogdan, this turned out to be much more work than I expected :smiley:. I'm not sure I've included `qtsvg` correctly  and I haven't made the phonon-backend-vlc` changes yet, will do those a little later.

**Comment by vaibhavsagar** [comment]
Oops, I did a bad rebase and overwrote some of my changes :frowning_face:

**Comment by vaibhavsagar** [comment]
Running `result/bin/vlc-wrapper --reset-config --reset-plugins-cache` gives me a working VLC!

**Comment by matthewbauer** [comment]
Related: #39037

**Comment by matthewbauer** [comment]
@peterhoeg does this branch work for you?

**Comment by peterhoeg** [comment]
Hey @vaibhavsagar, any chance you can incorporate the changes from #39037 as well? That PR is working fine except for the stale plugins spam on the console and it fixes a few other things (drops the npapi plugin, fixes live555, drops qt4)

**Comment by vaibhavsagar** [comment]
Sure!

**Comment by vaibhavsagar** [comment]
Hi @peterhoeg this is essentially your PR with a `phonon-backend-vlc` bump as well. Would you rather cherry-pick just that change to your PR?

**Comment by peterhoeg** [comment]
Eval breaks due to the missing npapi plugin. Do you feel like taking a stab at that? Would be great with ```nox-review wip``` as well.

**Comment by vaibhavsagar** [comment]
I'm not sure how to use `nox-review wip`, but I removed Firefox's dependency on `vlc_npapi`.

**Comment by vaibhavsagar** [comment]
@peterhoeg anything else you need from me?

**Comment by peterhoeg** [comment]
It doesn't build due to the wayland bits - I'm doing a bit of clean-up and pushing that your branch. Will also need @ttuegel's input later as it touches phonon and thus KDE.

**Comment by manveru** [comment]
Any chance you could add `--enable-chromecast` and `--enable-microdns` to allow using a chromecast? Although the microdns one is optional, it just allows automatic discovery, but requires a library that's not yet packaged.

**Comment by peterhoeg** [comment]
> requires a library that's not yet packaged.

Which one?

**Comment by peterhoeg** [comment]
@vaibhavsagar, I don't want to force push to your branch in order to fix the merge conflict - can you rebase against master please?

**Comment by manveru** [comment]
@peterhoeg https://github.com/videolabs/libmicrodns

**Comment by vaibhavsagar** [comment]
Rebased :smiley:.

**Comment by peterhoeg** [comment]
@ttuegel are you OK with dropping the qt4 version of phonon?

**Comment by peterhoeg** [comment]
vlc builds fine but the firefox wrapper eval is broken. Odd. I'll take a look later.

**Comment by peterhoeg** [comment]
@vaibhavsagar, looks like your rebase went a little off. You need to add ```uget-integrator``` to line 9.

**Comment by globin** [comment]
I just rebased this onto staging due to wayland being bumped there and vlc building with it, I'd be happy to merge, AFAICS everything is working fine.

**Comment by peterhoeg** [comment]
LGTM!

File: pkgs/applications/video/vlc/default.nix [hunk 1]
```diff
@@ -1,86 +1,81 @@
...
-    sed -e "s@/bin/echo@echo@g" -i configure
-  '' + optionalString withQt5 ''
-    # Make sure we only *add* "-std=c++11" to CXXFLAGS instead of overriding the
-    # values figured out by configure (for example "-g -O2").
-    sed -i -re '/^ *CXXFLAGS=("[^$"]+")? *$/s/CXXFLAGS="?/&-std=c++11 /' \
-      configure
-  '';
+  # vlc depends on a c11-gcc wrapper script which we don't have so we need to
+  # set the path to the compiler
+  BUILDCC = "${stdenv.cc}/bin/gcc";
```
Comment (lheckemann) [inline pkgs/applications/video/vlc/default.nix]: This should probably be `buildPackages.stdenv.cc` in order for cross-compilation to work.

**Pull request merged**